lib/components/app/session-timeout.tsx

```diff
@@ -0,0 +1,124 @@
+/* eslint-disable @typescript-eslint/ban-ts-comment */
+/* eslint-disable react/prop-types */
+import { Button, Modal } from 'react-bootstrap'
+import { connect } from 'react-redux'
+import { FormattedMessage } from 'react-intl'
+import React, { Component } from 'react'
+
+import * as uiActions from '../../actions/ui'
+
+/**
+ * This component makes the current session timeout
+ * by displaying a timeout warning one minute before the timeout,
+ * and by reloading the initial URL if there is no user-initiated
+ * actions within the timeout window.
+ */
+class SessionTimeout extends Component {
+  state = {
+    showTimeoutWarning: false,
+    timeoutObject: null,
+    timeoutStartMillis: 0
+  }
+
+  componentDidMount() {
+    // Wait one second or so after loading before probing for changes
+    // so that initialization actions can complete.
+    setTimeout(this.handleAfterInitialActions, 1500)
```

Comment:
What would happen on a slower computer/connection, if things aren't finished in 1.5 seconds?

Reply:
I guess the risk is that if some Redux actions get set off after this, then the `lastActionMillis` will be greater than `timeoutStartMillis` which would mean the page could time out without a user ever interacting with it?

Reply:
This just delays the start of the timer and the timeout check under `this.handleTimeoutWatch`, so that the `lastActionMills` state doesn't get unnecessarily updated with all the stuff that normally occurs during a page load. I'll add a comment about it.